interfaces/machine_parameters/ring.py

```diff
@@ -0,0 +1,604 @@
+
+        # Warning if kwargs were unused
+        if len(kwargs) > 0:
+            warnings.warn(
+                "Unused kwargs have been detected, " +
+                f"they are {list(kwargs.keys())}")
+
+    @classmethod
+    def direct_input(cls, Particle, length, alpha_0,
+                     momentum=None, kin_energy=None, energy=None,
+                     bending_field=None, bending_radius=None, orbit_bump=None,
+                     alpha_1=None, **kwargs):
+
+        # Getting the number of sections for the length list
+        length = np.array(length, ndmin=1, dtype=float)
+        n_sections = len(length)
+
+        # Checking that at least one synchronous data input is passed
+        syncDataTypes = ('momentum', 'kinetic_energy', 'total_energy',
+                         'bending_field')
+        syncDataInput = (momentum, kin_energy, energy, bending_field)
+
+        sync_data = {k: v for k, v in zip(syncDataTypes, syncDataInput)}
+        for k in sync_data:
+            if sync_data[k] is None:
+                sync_data[k] = [None] * n_sections
+            else:
+                if not isinstance(
+                        sync_data[k], ring_programs._synchronous_data_program):
+                    sync_data[k] \
+                        = getattr(ring_programs, k + '_program')(sync_data[k])
+
+        # Extending bending_radius to list of None if not defined
+        if orbit_bump is None:
+            bending_radius = [None] * n_sections
+
+        # Getting the momentum compaction factors (linear and non linear)
+        alphas = {'alpha_0': alpha_0, 'alpha_1': alpha_1,
+                  **{k: v for k, v in kwargs if 'alpha_' in k}}
+        for k in alphas:
+            if alphas[k] is None:
+                alphas[k] = [None] * n_sections
+            else:
+                if not isinstance(
+                        alphas[k], ring_programs.momentum_compaction):
+                    alphas[k] = ring_programs.momentum_compaction(
+                        alphas[k], order=int(k.split('_')[-1]))
+
+        # Extending orbit_bump to list of None if not defined
+        if orbit_bump is None:
+            orbit_bump = [None] * n_sections
+        else:
+            if not isinstance(orbit_bump, ring_programs.orbit_length_program):
+                orbit_bump = ring_programs.orbit_length_program(orbit_bump)
+
+        # Checking the data before building sections
+        if not all([len(v) == n_sections
+                    for v in {**sync_data, **alphas}.values()]):
+            raise excpt.InputDataError(
+                "Inconsistent data for Ring.direct_input !")
+
+        # Building all sections
+        RingSection_list = []
+        for i in range(n_sections):
+
+            sectionInput = {'length': length[i],
+                            'bending_radius': bending_radius,
+                            **{k: sync_data[k][i] for k in sync_data},
+                            **{k: alphas[k][i] for k in alphas},
```

Comment:
Short term solution to the casting issue:
```suggestion
                            **{k: sync_data[k][i].view(np.ndarray) for k in sync_data},
                            **{k: alphas[k][i].view(np.ndarray) for k in alphas},
```

Reply:
Included, I just added an if statement for None cases (keeping the inline coding for maximum compression rate 😄 )

Reply:
Hard approve!!!